File: 33.search-in-rotated-sorted-array.java
/*
 * @lc app=leetcode id=33 lang=java
 *
 * [33] Search in Rotated Sorted Array
 */

// @lc code=start
class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            int mid = (left + right) / 2;

            if ((nums[mid] < nums[0]) == (target < nums[0])) {
                if (target < nums[mid]) {
                    right = mid - 1;
                } else if (target > nums[mid]) {
                    left = mid + 1;
                } else {
                    return mid;
                }
            } else if (target < nums[0]) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1;
    }
}
// @lc code=end

File: 33.search-in-rotated-sorted-array.py
#
# @lc app=leetcode id=33 lang=python3
#
# [33] Search in Rotated Sorted Array
#

# @lc code=start
class Solution:
    def search(self, nums: List[int], target: int) -> int:

        left = 0
        right = len(nums) - 1

        while left <= right:
            mid = (left + right) // 2
            mid_val = nums[mid]

            if (nums[mid] < nums[0]) == (target < nums[0]):
                mid_val = nums[mid]
            
            else:
                if target < nums[0]:
                    mid_val = -float('inf')
                else:
                    mid_val = float('inf')
            
            if mid_val == target:
                return mid
            elif mid_val > target:
                right = mid - 1
            else:
                left = mid + 1
        
        return -1
            
# @lc code=end

